feat(http): add support for ArrayBuffer

cassand opened the pull request:
- Add the buffer option to store response in ArrayBuffer.
- Improve the interface to get back response independently of the buffer type.

Usage : Indicate the type of buffer to use in the request option

```
http.get('https://www.google.com', new RequestOptions({
    buffer: ResponseBuffer.ArrayBuffer
}));
```

At the moment there are 3 types of buffer available :
- ArrayBuffer
- Json
- Text

When the buffer is set to JSON the conversion will be done by the browser and therefore is non blocking for the UI thread (this was an [issue](https://github.com/angular/http/issues/79)).

All types of response will be equivalent regardless of the buffer type, only performance will be impacted (except for malformed JSON)

cassand committed:
feat(http): add support for ArrayBuffer

Add the buffer option to store response in ArrayBuffer
Improve the interface to get back response independently of the buffer type

zoechi commented:
Wasn't the `Http` development moved over to the https://github.com/angular/http repo?

cassand commented:
I heard about it, but the repo seems inactive, have you an idea of where I could make that PR ?

ericmartinezr commented:
cc @robwormald

jeffbcross commented:
@zoechi we've only moved issues over to angular/http for now, haven't moved code.

leosvelperez commented:
is there any plan to include this PR or support for ArrayBuffer and Blob any time soon?

mattetti commented:
As someone with the need for `ArrayBuffer` response, I would also very much appreciate feedback on this PR, especially because some APIs are being changed. /cc @jeffbcross

mhevery commented:
I think this is a duplicate of already landed feature: #7310 Closing. Speak up if this was done by mistake

cassand commented:
This feature diverges from #7310, that implements arguments for the request, whereas this one implements buffer options for the response. To confirm, you can take a look at [static_response.ts](https://github.com/angular/angular/blob/master/modules/%40angular/http/src/static_response.ts) where methods for arraybuffer and blob are still unimplemented.

However both the request and the response share the same 'body' logic, it might be good to refactor it.

mhevery commented:
@cassand Could you rebase and refactor? And I will work with you to get this merged in.

CaptainCodeman commented:
Wasn't it being rewritten ...

https://github.com/robwormald/http/tree/rewrite

cassand commented:
@mhevery Sure, thank you.

cassand committed:
refactor(http): share 'body' logic between Request and Response

mhevery commented:
@cassand Are you still planning on rebasing this or should we close?

cassand commented:
@mhevery The rebase is done, however, I encounter an issue with the public api and the second commit (refactor) :
Both the request and the response are exposed in the public api, and both inherit from 'RequestBody', which has methods that should be exposed, but can't because the public api guard does not look up for a parent class.
The first commit is safe to merge and the second need a validation for the public api.

cassand commented:
@mhevery I cleaned it and moved the changes to the tools for the public api in a separate commit, it should be fine to merge, if you need a change let me know

vicb commented on modules/@angular/http/src/backends/xhr_backend.ts:
```diff
@@ -41,7 +42,7 @@ export class XHRConnection implements Connection {
       }
       // load event handler
-      let onLoad = () => {
+      let onLoad = () => { 
```
revert

vicb commented on modules/@angular/http/http.ts:
```diff
@@ -29,6 +29,7 @@ export {Response} from './src/static_response';
 export {URLSearchParams} from './src/url_search_params';
 
 
+
```
revert

vicb commented on modules/@angular/http/src/request_body.ts:
```diff
@@ -0,0 +1,60 @@
+   * Returns the body as a string, presuming `toString()` can be called on the response body.
+   */
+  text(): string {
+    if (this._body instanceof URLSearchParams)
```
```
if (...) {
  return 1;
}

if (...) {
  return 2;
}

vicb commented on modules/@angular/http/src/static_request.ts:
```diff
@@ -134,6 +134,16 @@ export class Request extends RequestBody {
     }
   }
+
+  /* Override 'request_body' methods to ensure public api correctness */
```
Could you give more details on why you need those ? Is there a GH issue ?

cassand replied:
Indeed, it does not look up for methods in the parent class, therefore parents' methods are public and not listed in public_api_spec.ts

vicb replied:
Have you created an issue on ts-api-guardian ?

vicb replied:
Seems like it will be fixed via https://github.com/angular/ts-api-guardian/pull/6 thanks to @jasonchoimtt.

Please remove this useless methods and update the api spec accordingly for now.

vicb commented on modules/@angular/http/src/request_body.ts:
```diff
@@ -0,0 +1,75 @@
+      return new Blob([this._body]);
+    }
+
+    throw 'The request body isn\'t either a blob or an array buffer';
```
`throw new Error(...);` -> preserve the stack trace

vicb commented on modules/@angular/http/src/backends/xhr_backend.ts:
```diff
@@ -16,6 +16,7 @@ import {BrowserXhr} from './browser_xhr';
 
 const XSSI_PREFIX = /^\)\]\}',?\n/;
 
+
```
revert

vicb commented on modules/@angular/http/src/static_request.ts:
```diff
@@ -1,12 +1,14 @@
 import {URLSearchParams} from './url_search_params';
 
 
+
```
revert

vicb commented:
please the the minor comments

vicb commented:
Please ping us whenever you push more commits - I hadn't noticed the changes. Will take a look today. Thanks!

vicb commented on modules/@angular/http/src/backends/xhr_backend.ts:
```diff
@@ -98,6 +98,21 @@ export class XHRConnection implements Connection {
+            break;
+          case ResponseBuffer.Text:
+            _xhr.responseType = 'text';
+            break;
```
I think we enforce `default` now.
Could you please rebase on master ?

vicb commented on modules/@angular/http/src/base_response_options.ts:
```diff
@@ -34,11 +34,11 @@ import {ResponseOptionsArgs} from './interfaces';
+   * String, Object, ArrayBuffer or Blob representing the body of the {@link Response}.
    */
-  body: string|Object;
+  body: any;
```
why not union type here ?
consider creating a type alias if needed

vicb commented on modules/@angular/http/src/request_body.ts:
```diff
@@ -0,0 +1,75 @@
+ * https://fetch.spec.whatwg.org/#body
+ */
+export class RequestBody {
+  /** Body of the request **/
```
remove this comment

vicb commented on modules/@angular/http/src/request_body.ts:
```diff
@@ -0,0 +1,75 @@
+  /**
+   * Attempts to return body as parsed `JSON` object, or raises an exception.
+   */
+  json(): any {
```
what about `toJson`, `toText`, ... (similar to toString)

vicb replied:
oops those already exists... please discard my comment

vicb commented on modules/@angular/http/src/request_body.ts:
```diff
@@ -0,0 +1,75 @@
+ * Interface to extract request body
+ * https://fetch.spec.whatwg.org/#body
+ */
+export class RequestBody {
```
Could you finda better name `Response extends RequestBody` sounds weird

vicb commented:
Could you please address the comments.

cassand commented:
@vicb Sure, it is rebased and comments are addressed

vicb commented on modules/@angular/http/src/backends/xhr_backend.ts:
```diff
@@ -106,6 +106,21 @@ export class XHRConnection implements Connection {
+          case ResponseBuffer.Json:
+            _xhr.responseType = 'json';
+            break;
+          default:
```
```
default:
case ResponseBuffer.Text: (or whatever the name is)
  ...
```

`

vicb commented on modules/@angular/http/src/base_response_options.ts:
```diff
@@ -42,11 +42,11 @@ import {ResponseOptionsArgs} from './interfaces';
  */
 export class ResponseOptions {
-  // TODO: ArrayBuffer | FormData | Blob
+  // TODO: FormData | Blob
```
why Blob ?

cassand replied:
I had originally planned to support blob, it need about 10 lines, should I add it to this PR or implement the support for blob later ?

vicb replied:
you can do it later, will merge the PR now

spock123 replied:
Too bad, we have to resolve to using raw XHR to get binary data like pdfs etc... will longingly await Blob support ;)

vicb commented on modules/@angular/http/src/base_response_options.ts:
```diff
@@ -42,11 +42,11 @@ import {ResponseOptionsArgs} from './interfaces';
+   * String, Object, ArrayBuffer or Blob representing the body of the {@link Response}.
    */
-  body: string|Object;
+  body: string|Object|ArrayBuffer;
```
is the comment or the code right ?

vicb commented on modules/@angular/http/src/body.ts:
```diff
@@ -0,0 +1,83 @@
+ * https://fetch.spec.whatwg.org/#body
+ */
+export class Body {
+  /** Content of the request or the response **/
```
remove this commnent

vicb commented on modules/@angular/http/src/body.ts:
```diff
@@ -0,0 +1,83 @@
+
+
+/**
+ * Interface to extract request body, used by both {@link Request} and {@link Response}
```
`HTTP Request body used by both {@link Request} and {@link Response}`

vicb commented on modules/@angular/http/src/interfaces.ts:
```diff
@@ -54,8 +55,9 @@ export interface RequestArgs extends RequestOptionsArgs { url: string; }
 export type ResponseOptionsArgs = {
-  // TODO: Support Blob, ArrayBuffer, JSON
-  body?: string | Object | FormData; status?: number; statusText?: string; headers?: Headers;
+  // TODO: Support Blob, JSON
```
Is this still valid

vicb commented on modules/@angular/http/src/body.ts:
```diff
@@ -0,0 +1,83 @@
+ * Interface to extract request body, used by both {@link Request} and {@link Response}
+ * https://fetch.spec.whatwg.org/#body
+ */
+export class Body {
```
abstract class ?

vicb commented:
a few more comments. Sorry, you have to rebase again. Thanks.

vicb commented:
@robwormald or @jeffbcross could you please double check

vicb commented on modules/@angular/http/src/enums.ts:
```diff
@@ -57,3 +57,12 @@ export enum ContentType {
+/*
+ * Define which buffer to use to store the response
+ */
+export enum ResponseBuffer {
```
why can't we use `ContentType` (see right above) ?

cassand replied:
Some type such as `FormData` can be sent, but can't be received receive from a request, so we can't use `ContentType`

jeffbcross replied:
I'd rather change the name of `ContentType` to `RequestContentType` and rename `ResponseBuffer` to `ResponseContentType` to be more symmetrical. And set `type ContentType = RequestContentType` so it wouldn't be a breaking change. What do you think @cassand?

cassand replied:
Indeed, as the attribute` buffer` will be renamed to `responseType`, `ResponseContentType` will be more appropriate than `ResponseBuffer`

vicb commented on modules/@angular/http/src/base_request_options.ts:
```diff
@@ -67,7 +67,13 @@ export class RequestOptions {
+  /*
+   * Select a buffer to store the response, such as ArrayBuffer, Blob, Json (or Document)
+   */
+  buffer: ResponseBuffer;
```
could you please rename buffer to responseType

cassand replied:
I can rename buffer to `responseType` although I can't rename `ResponseBuffer` to `ResponseType` as it is already used in the [fetch specification](https://fetch.spec.whatwg.org/#responsetype) and has a different meaning :` "basic", "cors", "default"`

vicb commented on modules/@angular/http/src/backends/xhr_backend.ts:
```diff
@@ -106,6 +106,22 @@ export class XHRConnection implements Connection {
+          case ResponseBuffer.Json:
+            _xhr.responseType = 'json';
+            break;
+          default:
```
throw on default, ie "not supported"

jeffbcross commented:
I'm taking a look!

jeffbcross commented on modules/@angular/http/src/body.ts:
```diff
@@ -0,0 +1,84 @@
+ * HTTP request body used by both {@link Request} and {@link Response}
+ * https://fetch.spec.whatwg.org/#body
+ */
+export abstract class Body {
```
Nice!

jeffbcross commented on modules/@angular/http/src/backends/xhr_backend.ts:
```diff
@@ -106,6 +106,22 @@ export class XHRConnection implements Connection {
 
+      // Select the correct buffer type to store the response
+      if (isPresent(req.buffer) && isPresent(_xhr.responseType)) {
+        switch (req.buffer) {
```
@cassand san you help me understand why `isPresent(_xhr.responseType)` is here? I don't see where this would have been set prior to this statement.

cassand replied:
It helps detect if this header is supported by the browser, [the specification](https://www.w3.org/TR/XMLHttpRequest/#the-responsetype-attribute) states that the browser should set it to an empty string, therefore browsers that do no support it (IE9) will have a null value instead.

jeffbcross replied:
Or are there browsers that do not support `responseType`?

jeffbcross commented on modules/@angular/http/src/http_utils.ts:
```diff
@@ -38,4 +38,12 @@ export function getResponseURL(xhr: any): string {
+    view[i] = input.charCodeAt(i);
+  }
+  return view.buffer;
+}
```
@mprobst would you give a quick look at this fn?

mprobst replied:
Seems legit, any particular questions?

jeffbcross commented on tools/public_api_guard/http/index.d.ts:
```diff
@@ -147,15 +146,17 @@ export declare class Response {
 }
 
+export declare enum ResponseBuffer {
+    ArrayBuffer = 0,
```
The values don't need to be explicitly set.

jeffbcross commented:
LGTM once these are complete:
 - [ ] @mprobst gives thumbs up on `stringToArrayBuffer`, 
 - [ ] `buffer` property of `RequestOptions` is renamed to `responseType`
 - [ ] `ResponseBuffer` and `ContentType` renamed as described int the comment.

mprobst commented:
What I wonder about is how this works with XSSI prefixes, i.e. stripping the magic `']}),\n'` from the returned response. Parsing is done by the browser here, so how would we make this work together? Maybe @rjamet knows?

jeffbcross commented:
> What I wonder about is how this works with XSSI prefixes, i.e. stripping the magic ']}),\n' from the returned response. Parsing is done by the browser here, so how would we make this work together? Maybe @rjamet knows?

Hmm, good question. For Angular 1, we only strip if the response is a string: https://github.com/angular/angular.js/blob/99223a027809094ccb14fef2dc78199244943f97/src/ng/http.js#L135

jeffbcross commented:
Moving to RC.5 since I didn't realize commits had been pushed. 

@cassand is the PR ready for review? In the future, it helps if you comment with a mention if you push commits that are ready for review so that we'll get a notification.

cassand commented:
@jeffbcross It's almost ready, I just need to add a type alias for `RequestContentType`, the rest is ready for review

vicb commented:
what's the status here ?

cassand commented:
I can't figure out a proper way to export an alias to an enum for `ContentType`, as it is only used internally, this shouldn't be an issue. Otherwise, it is ready for review

jeffbcross commented:
I'll take another look today!

jeffbcross commented:
Merged at e7a8e2757b06d572f614f53b648d2fd75df370d2!
